samples/shared/AudioEngine.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SAMPLES_AUDIO_ENGINE_H
+#define SAMPLES_AUDIO_ENGINE_H
+
+
+#include <oboe/Oboe.h>
+
+
+class DefaultAudioStreamCallback;
+
+class AudioEngine {
+public:
+
+   template <class callback_ptr>
+    AudioEngine(std::shared_ptr<callback_ptr> callback) {
+        mCallback = std::dynamic_pointer_cast<oboe::AudioStreamCallback>(callback);
```

Comment:
Both megadrone and hello-oboe use instances of `AudioStreamCallback` - why is this cast necessary? If it's not necessary then this class no longer needs to be templated and `callback_ptr` can be replaced with `oboe::AudioStreamCallback`